### PR TITLE
apply stack optimizations to Alias method

### PR DESCRIPTION
**Pull request by wilzbach**
as suggested by @9il in #259

### CODE REVIEW

File: source/mir/random/discrete.d [hunk 1]
```diff
@@ -103,14 +103,15 @@ struct Discrete(T)
...
         import std.experimental.allocator : dispose, makeArray;
         auto alloc = Mallocator.instance;
 
-        auto small = alloc.makeArray!ProbsIndexed(n); // columns that need to be filled
-        auto large = alloc.makeArray!ProbsIndexed(n); // used to fill columns
+        auto stack = alloc.makeArray!ProbsIndexed(n);
+        // front: columns that need to be filled
+        size_t smCounter;
+        // back: used to fill columns
+        size_t lgCounter = n - 1;
```
Comment (9il) [inline source/mir/random/discrete.d:114]: `size_t lgCounter = n;` this will make some code more simple

File: source/mir/random/discrete.d [hunk 2]
```diff
@@ -119,16 +120,16 @@ struct Discrete(T)
             // 1 is the average probability and depending on the ratio, we either
             // need to add values to a block or remove values
             if (sp < 1)
-                small[smCounter++] = ProbsIndexed(sp, i);
+                stack[smCounter++] = ProbsIndexed(sp, i);
             else
-                large[lgCounter++] = ProbsIndexed(sp, i);
+                stack[lgCounter--] = ProbsIndexed(sp, i);
```
Comment (9il) [inline source/mir/random/discrete.d:125]: ... so, `--lgCounter`

File: source/mir/random/discrete.d [hunk 3]
```diff
@@ -119,16 +120,16 @@ struct Discrete(T)
...
         }
 
         // as long as there are elements in both stacks
         while (smCounter > 0 && lgCounter > 0)
```
Comment (9il) [inline source/mir/random/discrete.d]: `lgCounter < n`

File: source/mir/random/discrete.d [hunk 4]
```diff
@@ -119,16 +120,16 @@ struct Discrete(T)
...
         {
-            auto sm = small[--smCounter];
-            auto lg = large[--lgCounter];
+            auto sm = stack[--smCounter];
+            auto lg = stack[++lgCounter];
```
Comment (9il) [inline source/mir/random/discrete.d:132]: `lgCounter++`

Comment (9il) [inline source/mir/random/discrete.d:132]: and etc.

**Comment by 9il** [comment]
sorry I can be really wimp. `n-1` / `n` change is not required. But loop condition should be fixed

**Comment by wilzbach** [comment]
> sorry I can be really wimp. n-1 / n change is not required. But loop condition should be fixed

No worries. Should be done & including your other suggestions ;-)

**Pull request merged**